### PR TITLE
Only access loss tensor every logging_steps

### PR DESCRIPTION
**Pull request by jysohn23**
* tensor.item() was being called every step. This must not be done for XLA:TPU tensors as it's terrible for performance causing TPU<>CPU communication at each step. On RoBERTa MLM for example, it reduces step time by 30%, should be larger for smaller step time models/tasks.
* Train batch size was not correct in case a user uses the `per_gpu_train_batch_size` flag
* Avg reduce loss accross eval shards
* Log TPU debug metrics before last epoch break

### CODE REVIEW

**Review by julien-c — APPROVED** [review]

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -631,7 +631,7 @@ def train(self, model_path: Optional[str] = None, trial: Union["optuna.Trial", D
         logger.info("***** Running training *****")
         logger.info("  Num examples = %d", self.num_examples(train_dataloader))
         logger.info("  Num Epochs = %d", num_train_epochs)
-        logger.info("  Instantaneous batch size per device = %d", self.args.per_device_train_batch_size)
+        logger.info("  Instantaneous batch size per device = %d", train_dataloader.batch_size)
```
Comment (JetRunner) [inline src/transformers/trainer.py:634]: Why is this change necessary? The prompt is `Instantaneous batch size per device`. Also, this change seems to create inconsistency for logging.

Comment (jysohn23) [reply src/transformers/trainer.py:634]: 1. If using `--per_gpu_train_batch_size` (legacy flag for `per_device_train_batch_size`), the argument is straight up incorrect.
2. That batch size is per device.

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @jysohn23

**Comment by LysandreJik** [comment]
Style issue will be solved with merge @sgugger

File: src/transformers/trainer.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/trainer.py:634]: No the batch size is not per device. It is on TPUs but not if you a user has a `DataParallel` model.

**Review by sgugger — APPROVED** [review]
Thanks for fixing this! I'd remove the first change in the logs though.

**Comment by jysohn23** [comment]
> Thanks for fixing this! I'd remove the first change in the logs though.

Thanks for the review! Done.
